data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetStateByBlockRoot.java

```diff
@@ -0,0 +1,104 @@
+
+import io.javalin.core.util.Header;
+import io.javalin.http.Context;
+import io.javalin.http.Handler;
+import io.javalin.plugin.openapi.annotations.HttpMethod;
+import io.javalin.plugin.openapi.annotations.OpenApi;
+import io.javalin.plugin.openapi.annotations.OpenApiContent;
+import io.javalin.plugin.openapi.annotations.OpenApiParam;
+import io.javalin.plugin.openapi.annotations.OpenApiResponse;
+import java.util.Map;
+import java.util.Optional;
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.teku.api.ChainDataProvider;
+import tech.pegasys.teku.api.DataProvider;
+import tech.pegasys.teku.api.response.StateSszResponse;
+import tech.pegasys.teku.beaconrestapi.schema.BadRequest;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.provider.JsonProvider;
+
+public class GetStateByBlockRoot implements Handler {
+  public static final String ROUTE = "/teku/v1/beacon/blocks/:block_id/state";
+  private final ChainDataProvider chainDataProvider;
+  private final JsonProvider jsonProvider;
+
+  public GetStateByBlockRoot(final DataProvider dataProvider, final JsonProvider jsonProvider) {
+    this(dataProvider.getChainDataProvider(), jsonProvider);
+  }
+
+  public GetStateByBlockRoot(
+      final ChainDataProvider chainDataProvider, final JsonProvider jsonProvider) {
+    this.jsonProvider = jsonProvider;
+    this.chainDataProvider = chainDataProvider;
+  }
+
+  @OpenApi(
+      path = ROUTE,
+      method = HttpMethod.GET,
+      summary = "Get SSZ State By Block id",
+      tags = {TAG_TEKU},
+      description = "Download the state SSZ object for given identifier- by block root, keyword, or slot.",
+      pathParams = {@OpenApiParam(name = PARAM_BLOCK_ID, description = PARAM_BLOCK_ID_DESCRIPTION)},
+      responses = {
+        @OpenApiResponse(
+            status = RES_OK,
+            content = @OpenApiContent(type = "application/octet-stream")),
+        @OpenApiResponse(status = RES_BAD_REQUEST),
+        @OpenApiResponse(status = RES_NOT_FOUND),
+        @OpenApiResponse(status = RES_INTERNAL_ERROR),
+        @OpenApiResponse(status = RES_SERVICE_UNAVAILABLE, description = SERVICE_UNAVAILABLE)
+      })
+  @Override
+  public void handle(@NotNull final Context ctx) throws Exception {
+    final Map<String, String> pathParamMap = ctx.pathParamMap();
+    ctx.header(Header.CACHE_CONTROL, CACHE_NONE);
+
+    SafeFuture<Optional<StateSszResponse>> future =
+        chainDataProvider.getBeaconStateSszByBlockRoot(pathParamMap.get(PARAM_BLOCK_ID));
+    ctx.result(
+        future.thenApplyChecked(
+            result -> {
+              if (result.isEmpty()) {
+                ctx.status(SC_NOT_FOUND);
+                ctx.result(
+                    BadRequest.serialize(
+                        jsonProvider,
+                        SC_NOT_FOUND,
+                        "State by block root not found: " + pathParamMap.get(PARAM_BLOCK_ID)));
```

Comment:
Feels odd to be calling ctx.result twice rather than just returning the serialised JSON.  You'd have to make it `future.<Object>thenApplyChecked` above I believe.

Though the `result` javadoc does say it overwrites the existing result so I guess it works...

Reply:
yep it should just return the string. fixed